### PR TITLE
[v2.7] End to End test Code Coverage with Jenkins pipeline and setup

### PR DESCRIPTION
**Pull request by igomez06**
This PR includes code coverage custom binaries for both rancher and agent. As well as, Dockerfiles for both rancher and rancher-agent. Setup scripts for creating the binaries and the docker images. Scripts for setting the Jenkins environment to have a test coverage run for all our tests. New extensions that kill the containers that are running the rancher/agent images to have the coverage reports written, and code to retreive those reports from their pods.

## Issue: 
 https://github.com/rancher/rancher/issues/38282, https://github.com/rancher/qa-tasks/issues/520
 
## Problem
 Need to have a way to measure our code coverage of rancher/rancher from out end to end tests, including all of python tests (validation and integration), and our go automation tests.
 
## Solution

**Overview**
1. Custom test binaries for rancher and rancher-agent. This is so we can encapsulate the rancher/main.go and rancher/cmd/agent/main.go in `main` functions to be able run the binaries as tests to be able to use the `go build -cover` functionality but also be able to kill the bin to have reports written.
2. Custom Dockerfiles for both rancher and rancher-agent to be run in deployments of the local cluster and the downstream clusters provisioned by rancher. 
3. New extensions that work as kubectl exec calls to able to "kill" the container in the its respective pod that is running the custom image (rancher or rancher-agent) and also calls to be able to retrieve said coverage reports from the pods to a local destination.
4. New Jenkins pipeline that runs all of our end to end tests, retrieves the code coverage reports, and publishes the reports.
5. Set up code the sets up the environment for our aforementioned Jenkins pipeline, that uses corral to create a rancher instance, update the rancher instance to use the custom code coverage images, and provision clusters to be used in the end to end tests.
 
## Testing
 code coverage report Jenkins job

### CODE REVIEW

File: tests/framework/extensions/kubeconfig/exec.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+		Stdout: logStreamer,
+		Stderr: os.Stderr,
+	})
+	return logStreamer, err
+}
+
+// CopyFileFromPod is function that copies files from a pod. The parameter takes
+// the kubeconfig in form of a restclient.Config object, the pod name, the namespace of the pod, the filename, and then
+// the local destination (dest) where the file will be copied to.
+func CopyFileFromPod(restConfig *restclient.Config, clientConfig clientcmd.ClientConfig, podName, namespace, filename, dest string) error {
```
Comment (igomez06) [inline tests/framework/extensions/kubeconfig/exec.go:87]: @markusewalker @caliskanugur @izaac this PR is the same to v2.6 except for this function. release/v2.7 has an updated version of `"k8s.io/kubectl/pkg/cmd/cp"` which requires a slightly different configuration.

**Review by markusewalker — DISMISSED** [review]
LGTM

File: tests/framework/extensions/kubeconfig/restgetter.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+// ToDiscoveryClient returns discovery client
+func (r *RestGetter) ToDiscoveryClient() (discovery.CachedDiscoveryInterface, error) {
+	return nil, nil
+}
+
+// ToRESTMapper returns a restmapper
+func (r *RestGetter) ToRESTMapper() (meta.RESTMapper, error) {
+	return nil, nil
+}
```
[the remark this answers is not in the document]
Comment (igomez06) [reply tests/framework/extensions/kubeconfig/restgetter.go:37]: They are a base Interface requirement for the new way set up a copyOptions. I prefer the old way, but this end users more flexibility I guess.

File: tests/v2/codecoverage/Jenkinsfile_buildcodecoverage [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  if (job_name.contains('/')) { 
+    job_names = job_name.split('/')
+    job_name = job_names[job_names.size() - 1] 
+  }
+  def testContainer = "${job_name}${env.BUILD_NUMBER}_test"
+  def imageName = "rancher-code-coverage-image-build-${job_name}${env.BUILD_NUMBER}"
+  def buildDockerContainer = "build-code-cover-docker-container"
+  def envFile = ".env"
+  def rancherConfig = "rancher_env.config"
+  def branch = "release/v2.6"
```
[the remark this answers is not in the document]
Comment (igomez06) [reply tests/v2/codecoverage/Jenkinsfile_buildcodecoverage:14]: You right! v2.7 should be the default. I'll change it right now.

[the remark this answers is not in the document]
Comment (igomez06) [reply tests/v2/codecoverage/Jenkinsfile_buildcodecoverage]: Fixed

File: tests/framework/extensions/kubeconfig/restgetter.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (caliskanugur) [reply tests/framework/extensions/kubeconfig/restgetter.go:37]: Makes sense, thanks for the explanation

**Review by markusewalker — APPROVED** [review]

**Review by izaac — APPROVED** [review]

**Review by caliskanugur — DISMISSED** [review]

**Comment by igomez06** [comment]
Last update to account for drone failures, that just got merged in.

**Review by markusewalker — DISMISSED** [review]

**Review by markusewalker — APPROVED** [review]

**Review by caliskanugur — APPROVED** [review]

**Pull request merged**